payreq.js

```diff
@@ -487,24 +493,35 @@ const encode = (inputData, addDefaults) => {
   // calculate the smallest possible integer (removing zeroes) and add the best
   // multiplier (m = milli, u = micro, n = nano, p = pico)
   if (data.satoshis) {
-    let mSats = BigNumber(1000).mul(data.satoshis)
-    let mSatsString = mSats.toString(10).replace(/\.\d*$/, '')
+    if (!/^\d*(\.\d{0,4})?$/.test(data.satoshis)) {
+      throw new Error('Invalid satoshis value')
+    }
+
+    let satoshis4 = data.satoshis + '0000'
+    if (data.satoshis.indexOf('.') !== -1) {
+      satoshis4 = data.satoshis.replace(/\.(\d{0,4})$/, function (m, p1) {
+        return (p1 + '0000').slice(0, 4)
+      })
+    }
+
+    let mSats = new BN(satoshis4, 10)
+    let mSatsString = mSats.toString(10).slice(0, -1)
     let mSatsLength = mSatsString.length
     if (mSatsLength > 11 && mSatsString.slice(-11) === '00000000000') {
       multiplier = ''
-      value = mSats.div(1e11).toString(10)
+      value = divmod2str(mSats.divmod(MULTIPLIERS.p))
     } else if (mSatsLength > 8 && mSatsString.slice(-8) === '00000000') {
       multiplier = 'm'
-      value = mSats.div(1e8).toString(10)
+      value = divmod2str(mSats.divmod(MULTIPLIERS.n))
     } else if (mSatsLength > 5 && mSatsString.slice(-5) === '00000') {
       multiplier = 'u'
-      value = mSats.div(1e5).toString(10)
+      value = divmod2str(mSats.divmod(MULTIPLIERS.u))
     } else if (mSatsLength > 2 && mSatsString.slice(-2) === '00') {
       multiplier = 'n'
-      value = mSats.div(1e2).toString(10)
+      value = divmod2str(mSats.divmod(MULTIPLIERS.m))
     } else {
       multiplier = 'p'
-      value = mSats.mul(10).toString(10)
+      value = mSats.toString(10)
```

Comment:
Is any of this actually **necessary** though?
The underlying API shouldn't need it?

Reply:
sorry, this is even will not work..
TBH we need split decode/encode to small functions and test them, otherwise this will not ready for using, because this library working with money, not eth kitties